File: sdk/sdk-python/netledger_sdk/methods/account.py
"""Account management operations for the NetLedger API."""

from typing import Optional
from urllib.parse import quote

from ..http_client import HttpClient
from ..models import Account, AccountEnumerationQuery, EnumerationResult
from ..exceptions import NetLedgerValidationError


class AccountMethods:
    """Account management operations."""

    def __init__(self, client: HttpClient):
        """
        Initialize account methods.

        Args:
            client: The HTTP client to use.
        """
        self._client = client

    def create(self, name: str, notes: Optional[str] = None) -> Account:
        """
        Create a new account.

        Args:
            name: The account name.
            notes: Optional notes for the account.

        Returns:
            The created account.

        Raises:
            NetLedgerValidationError: If the name is empty.
            NetLedgerConnectionError: If unable to connect to the server.
            NetLedgerApiError: If the server returns an error.
        """
        if not name or not name.strip():
            raise NetLedgerValidationError('Account name cannot be empty', 'name')

        body = {'name': name}
        if notes:
            body['notes'] = notes

        response = self._client.put('/v1/accounts', body)
        if not response.data:
            raise ValueError('No data returned from server')
        return Account.from_dict(response.data)

    def get(self, account_guid: str) -> Account:
        """
        Get an account by GUID.

        Args:
            account_guid: The account GUID.

        Returns:
            The account.

        Raises:
            NetLedgerConnectionError: If unable to connect to the server.
            NetLedgerApiError: If the server returns an error (404 if not found).
        """
        response = self._client.get(f'/v1/accounts/{account_guid}')
        if not response.data:
            raise ValueError('No data returned from server')
        return Account.from_dict(response.data)

    def get_by_name(self, name: str) -> Account:
        """
        Get an account by name.

        Args:
            name: The account name.

        Returns:
            The account.

        Raises:
            NetLedgerValidationError: If the name is empty.
            NetLedgerConnectionError: If unable to connect to the server.
            NetLedgerApiError: If the server returns an error (404 if not found).
        """
        if not name or not name.strip():
            raise NetLedgerValidationError('Account name cannot be empty', 'name')

        encoded_name = quote(name, safe='')
        response = self._client.get(f'/v1/accounts/byname/{encoded_name}')
        if not response.data:
            raise ValueError('No data returned from server')
        return Account.from_dict(response.data)

    def exists(self, account_guid: str) -> bool:
        """
        Check if an account exists.

        Args:
            account_guid: The account GUID.

        Returns:
            True if the account exists.

        Raises:
            NetLedgerConnectionError: If unable to connect to the server.
        """
        return self._client.head(f'/v1/accounts/{account_guid}')

    def delete(self, account_guid: str) -> None:
        """
        Delete an account.

        Args:
            account_guid: The account GUID.

        Raises:
            NetLedgerConnectionError: If unable to connect to the server.
            NetLedgerApiError: If the server returns an error (404 if not found).
        """
        self._client.delete(f'/v1/accounts/{account_guid}')

    def enumerate(self, query: Optional[AccountEnumerationQuery] = None) -> EnumerationResult:
        """
        Enumerate accounts with optional filtering and pagination.

        Args:
            query: Query parameters for filtering and pagination.

        Returns:
            Enumeration result containing accounts.

        Raises:
            NetLedgerConnectionError: If unable to connect to the server.
            NetLedgerApiError: If the server returns an error.
        """
        if query is None:
            query = AccountEnumerationQuery()

        params = [f'maxResults={query.max_results}', f'skip={query.skip}']
        if query.search_term:
            params.append(f'searchTerm={quote(query.search_term, safe="")}')

        path = f'/v1/accounts?{"&".join(params)}'
        response = self._client.get(path)

        if not response.data:
            return EnumerationResult()

        return EnumerationResult.from_dict(response.data, Account.from_dict)

File: sdk/sdk-js/src/methods/account.ts
import { Account, AccountEnumerationQuery, EnumerationResult } from '../models';
import { HttpClient } from '../http-client';
import { NetLedgerValidationError } from '../errors';

/**
 * Account management operations.
 */
export class AccountMethods {
    private readonly client: HttpClient;

    constructor(client: HttpClient) {
        this.client = client;
    }

    /**
     * Create a new account.
     * @param name Account name.
     * @param notes Optional notes.
     * @returns The created account.
     */
    async create(name: string, notes?: string, labels?: string[], tags?: Record<string, string>): Promise<Account> {
        if (!name || name.trim() === '') {
            throw new NetLedgerValidationError('Account name cannot be empty', 'name');
        }
        const response = await this.client.put<Account>('/v1/accounts', { Name: name, Notes: notes, Labels: labels, Tags: tags });
        if (!response.Data) {
            throw new Error('No data returned from server');
        }
        return response.Data;
    }

    /**
     * Get an account by identifier.
     * @param accountId The account identifier.
     * @returns The account.
     */
    async get(accountId: string): Promise<Account> {
        const response = await this.client.get<Account>(`/v1/accounts/${accountId}`);
        if (!response.Data) {
            throw new Error('No data returned from server');
        }
        return response.Data;
    }

    /**
     * Get an account by name.
     * @param name The account name.
     * @returns The account.
     */
    async getByName(name: string): Promise<Account> {
        if (!name || name.trim() === '') {
            throw new NetLedgerValidationError('Account name cannot be empty', 'name');
        }
        const encodedName = encodeURIComponent(name);
        const response = await this.client.get<Account>(`/v1/accounts/byname/${encodedName}`);
        if (!response.Data) {
            throw new Error('No data returned from server');
        }
        return response.Data;
    }

    /**
     * Check if an account exists.
     * @param accountId The account identifier.
     * @returns True if the account exists.
     */
    async exists(accountId: string): Promise<boolean> {
        return await this.client.head(`/v1/accounts/${accountId}`);
    }

    /**
     * Delete an account.
     * @param accountId The account identifier.
     */
    async delete(accountId: string): Promise<void> {
        await this.client.delete(`/v1/accounts/${accountId}`);
    }

    /**
     * Enumerate accounts with optional filtering and pagination.
     * @param query Query parameters.
     * @returns Enumeration result containing accounts.
     */
    async enumerate(query?: AccountEnumerationQuery): Promise<EnumerationResult<Account>> {
        const params = new URLSearchParams();
        if (query) {
            if (query.MaxResults !== undefined) params.append('maxResults', query.MaxResults.toString());
            if (query.Skip !== undefined) params.append('skip', query.Skip.toString());
            if (query.SearchTerm) params.append('searchTerm', query.SearchTerm);
            if (query.Labels && query.Labels.length > 0) params.append('labels', query.Labels.join(','));
            if (query.Tags && Object.keys(query.Tags).length > 0) {
                params.append('tags', Object.entries(query.Tags).map(([key, value]) => `${key}=${value}`).join(','));
            }
        }
        const queryString = params.toString();
        const path = queryString ? `/v1/accounts?${queryString}` : '/v1/accounts';
        const response = await this.client.get<EnumerationResult<Account>>(path);
        return response.Data || { TotalRecords: 0, RecordsRemaining: 0, EndOfResults: true };
    }
}
